csrc/ops/composite.cpp

```diff
@@ -54,42 +54,71 @@ TensorView* dropout_backward(TensorView* dy, TensorView* mask, Val* scale) {
   return dx;
 }
 
-TensorView* linear(TensorView* a, TensorView* b, TensorView* bias) {
-  // TODO: Support 1+ dimensional A.
-  NVF_CHECK(
-      (a->nDims() == 2 && b->nDims() == 2),
-      "Only 2-D Inputs and Weights are currently supported in Linear!");
-
-  std::vector<bool> bcast_dims(a->nDims() + 1, false);
-  // A: [M, Bcast, K]
-  // B: [Bcast, N, K]
-  bcast_dims.at(bcast_dims.size() - 2) = true;
-  auto* tv0b = broadcast(a, bcast_dims);
-  bcast_dims.at(bcast_dims.size() - 2) = false;
-  bcast_dims.at(bcast_dims.size() - 3) = true;
-  auto* tv1b = broadcast(b, bcast_dims);
+namespace {
 
-  NVF_CHECK(
-      a->getDataType().value() == b->getDataType().value(),
-      "data types of inputs to matmul don't match");
-
-  auto* output = fusedMultiplySum(tv0b, tv1b, {-1});
-  if (bias) {
-    NVF_CHECK(
-        (bias->nDims() <= a->nDims()), "bias should be broadcastable to A");
-    NVF_CHECK(
-        a->getDataType().value() == bias->getDataType().value(),
-        "bias doesn't match input/weight dtype");
-    auto* bias_with_cast = maybeCastOp(output->getDataType().value(), bias);
-    auto* bcast_bias = ops::maybeBroadcast({output, bias_with_cast})[1];
-    auto* bias_output = add(output, bcast_bias);
-    return maybeCastOp(a->getDataType().value(), bias_output);
+static TensorView* newForLinear(
+    TensorView* tv_a,
+    TensorView* tv_b,
+    TensorView* bias) {
+  auto orig_domain_a =
+      TensorDomain::noReductions(tv_a->getMaybeRFactorDomain());
+  auto orig_domain_b =
+      TensorDomain::noReductions(tv_b->getMaybeRFactorDomain());
+
+  // Linear: a = {*, in_features}, b = {out_features, in_features} /
+  // {in_features} For the linear output, all but the last dimension are the
+  // same shape as the first input. The last dimension is out_features (if present).
+  auto ndims_out = (orig_domain_a.size() - 1) + (orig_domain_b.size() - 1);
+
+  const std::vector<IterDomain*>& mapping_a = ops::mapLinearOpIterDomains(
+      orig_domain_a, MatmulRole::INPUT_A, ndims_out);
+  const std::vector<IterDomain*>& mapping_b = ops::mapLinearOpIterDomains(
+      orig_domain_b, MatmulRole::INPUT_B, ndims_out);
+  std::vector<IterDomain*> mapping_bias (ndims_out, nullptr);
+  if (bias != nullptr){
+    auto bias_domain = TensorDomain::noReductions(bias->getMaybeRFactorDomain());
+    mapping_bias = ops::mapLinearOpIterDomains(bias_domain, MatmulRole::INPUT_C, ndims_out);
   }
-  return maybeCastOp(a->getDataType().value(), output);
+
+  std::vector<IterDomain*> out_domain = ops::newOutputDomain({mapping_a, mapping_b, mapping_bias});
```

Comment:
What do you mean by take precedence?
For each position, I am using all the corresponding iterdomains from all inputs to create the output iterdomain. You are correct though, that we can only use mapping_b and it will work.

However, if one is Symbolic and the other Concrete, we can resolve the extents for the output if we use both. Wdyt?